### PR TITLE
New tool to update ppa recipes (infra)

### PR DESCRIPTION
**Pull request by Hook25**
## Description

As a part of the effort of splitting the build process of debs into more single responsibility steps, this PR creates a new tool that splits the updating of the recipe from the script `lp-recipe-update-build.py`, making it more testable and moving the recipe to the repository (where previously it was only updated but the ground truth was in the PPA).

## Resolved issues

Resolves: https://warthogs.atlassian.net/browse/CHECKBOX-1027

## Documentation

This adds a new section at the top of the script that documents what its purpose is. The help prompt was reworded options were added and documented.

## Tests

This adds a new unit test file that covers all the additions. The utils file is tested in [this PR](https://github.com/canonical/checkbox/pull/854) that introduced the functions. Depending what PR lands first I will rebase and move the tests.

### CODE REVIEW

File: tools/release/lp_update_recipe.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Kicks off a recipe build for a branch in Launchpad which has an associated
+recipe.
+Meant to be used as part of a checkbox release to the Hardware Certification
```
Comment (kissiel) [inline tools/release/lp_update_recipe.py:24]: heh, I think this is no longer up to date

File: tools/release/lp_update_recipe.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+public PPA and the ~checkbox-dev testing PPA.
+
+NOTE: The Hardware Certification public PPA has no further use over the
```
Comment (kissiel) [inline tools/release/lp_update_recipe.py:27]: and this can be deleted too

File: tools/release/lp_update_recipe.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+"""
+
+import sys
+import textwrap
+import argparse
+
+from utils import get_build_recipe
+
+
+def get_build_path(recipe_name: str):
```
Comment (kissiel) [inline tools/release/lp_update_recipe.py:39]: `-> str` missing

File: tools/release/lp_update_recipe.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        component_name = "-".join(package[2:])  # 0-1 are checkbox provider
+        return f"providers/{component_name}"
+    return "-".join(package)
+
+
+def get_updated_build_recipe(
+    recipe_name: str,
+    version: str,
+    revision: str,
+):
```
Comment (kissiel) [inline tools/release/lp_update_recipe.py:52]: ditto

File: tools/release/utils.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    """return Launchpad(
+        None, None, None, service_root="production", version="devel"
+    )"""
+    credentials = os.getenv("LP_CREDENTIALS")
+    if not credentials:
+        raise SystemExit("LP_CREDENTIALS environment variable missing")
+
+    credentials = Credentials.from_string(credentials)
+    return Launchpad(
+        credentials, None, None, service_root="production", version="devel"
```
Comment (kissiel) [inline tools/release/utils.py:19]: aren't those `None`s unnecessary?

File: tools/release/utils.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    build_recipe = (
+        recipe for recipe in project.recipes if recipe.name == recipe_name
+    )
+    try:
+        return next(build_recipe)
+    except StopIteration:
+        # There is no recipe with that name
+        all_recipe_names = ", ".join(recipe.name for recipe in project.recipes)
+        raise SystemExit(
```
Comment (kissiel) [inline tools/release/utils.py:41]: raising exceptions while in an exc handling block is discouraged and iffy.

as an alternative you can collect the list comprehension, and run an if over it, like so:
```
build_recipe = [r for r in project.recipes if r.name = recipe_name]
if build_recipe:
    return build_recipe[0]
else:
    raise (...)
```

File: tools/release/utils.py (hunk 5 above)
Comment (kissiel) [reply tools/release/utils.py:19]: They are not unnecessary.
Ignore me here :)

File: tools/release/utils.py (hunk 6 above)
Comment (kissiel) [reply tools/release/utils.py:41]: Oh, that's smart! And it justifies the complexity. 
Could you copy over the comment you wrote (the second part) to the code? So the next person doesn't have to tilt their head while reading

Comment (Hook25) [reply tools/release/utils.py:45]: done

**Pull request merged**